py3status/modules/emerge_status.py

```diff
@@ -0,0 +1,160 @@
+        # if emerge is running and total != 0 show
+        '[[\?if=is_running [\?if=!total=0 {current}/{total}'
+        # show action and cat/pkg only if user wants it
+        '[\?if=show_action {action} ][\?if=show_pkg {category}/{pkg}]]]|'
+        # if no emerge is running but show_stopped is set show 0/0
+        '[\?if=!hide_if_stopped 0/0]]')
+    hide_if_stopped = False
+    prefix = "EMRG: "
+    show_action = True
+    show_pkg = True
+
+    def __init__(self):
+        self.ret_default = {
+            'current': 0,
+            'total': 0,
+            'category': "",
+            'pkg': "",
+            'action': "",
+            'is_running': False}
+
+    def _emergeRunning(self):
+        """
+        Check if emerge is running.
+        Returns true if at least one instance of emerge is running.
+        """
+        try:
+            self.py3.command_output(['pgrep', 'emerge'])
+            return True
+        except Exception:
+            return False
+
+    def post_config_hook(self):
+        if not self.py3.check_commands('emerge'):
+            raise Exception("emerge not installed")
+
+    def _getProgress(self):
+        """
+        Get current progress of emerge.
+
+        returns a dict containing current and total value.
+        """
+        emerge_log_file = '/var/log/emerge.log'
+
+        ret = {}
+
+        input_data = []
+
+        """
+        Try to open file, if unable throw error to be catched later
+        """
+        try:
+            with open(emerge_log_file, 'r') as fp:
+                input_data = fp.readlines()
+        except IOError as err:
+            raise Exception(err)
+
+        """
+        Traverse emerge.log from bottom up to get latest information
+        """
+
+        input_data.reverse()
+
+        for line in input_data:
+            if "*** terminating." in line:
+                # Copy content of ret_default no only the references
+                ret = copy.deepcopt(self.ret_default)
+                break
+            else:
+                status_re = re.compile(
+                    "\((?P<cu>[\d]+) of (?P<t>[\d]+)\) "
+                    "(?P<a>[a-zA-Z\/]+( [a-zA-Z]+)?) "
+                    "\((?P<ca>[\w\-]+)\/(?P<p>[\w\.]+)"
+                )
+                res = status_re.search(line)
+                if res is not None:
+                    ret['current'] = res.group('cu')
+                    ret['total'] = res.group('t')
+                    ret['category'] = res.group('ca')
+                    ret['pkg'] = res.group('p')
+                    ret['action'] = res.group('a')
+                    break
+        return ret
+
+    def emerge_status(self):
+        """
+        Emerge Status main routine
+        """
+        response = {}
+        response['cached_until'] = self.py3.time_in(self.cache_timeout)
+        response['is_running'] = self._emergeRunning()
+        ret = copy.deepcopy(self.ret_default)
```

Comment:
Is this too much? Does `ret = self.ret_default` works fine?

Reply:
Since ret_default is a dictionary filled with references the assignment `ret = self.ret_default` will copy all references and not their actual values. As a result when setting e.g. ret['current'] we override the default value as well. I wanted to use ** here but this has only been introduced in python 3.5 and travis for 3.4 did not like it.